Base pipeline

mumichae opened the pull request:
A basic end-to-end pipeline for extracting 3'UTR regions and computing MAPS score on a variant subset within those regions.

- snakemake pipeline
- README containing install, setup and pipeline calls
- config with input files and parameters
- precommit, githooks, lint configurations added

Commit:
added files

Commit:
added subsetting option for gnomad datatsets & removed coverage information

Commit:
added exploratory analysis of PolyA_DB

Commit:
added simple region extraction from PolyA_DB

Commit:
split up rules into different files, added data download for FASTA reference

Commit:
extract hexamers from PolyA_DB

Commit:
correct interval for hexamer on (-) strand

Commit:
added linters and reformatted code

Commit:
add conversion from bed to interval rule before running MAPS

Commit:
added gcp rule for MAPS

Commit:
added GCP MAPS script

Commit:
add basic stats file for PolyA_DB features

Commit:
added description of project to README

Commit:
use hail from gcp channel

Commit:
add local & GCP run information

Commit:
attempt to download PolyA_DB, added URLs for databases

Commit:
added dependency graphs

vladsavelyev commented:
Hi Michaela, I'm about to take a look at the codebase, but one thing comes to mind right now: would you be able to add some info and/or pointers on what is the MAPS score, and why do we need to calculate it? Also, same on PAS hexamers. Would be helpful to add it into the README and give some clarity what's the challenge the code is solving.

vladsavelyev commented on .flake8:
```diff
@@ -0,0 +1,19 @@
+docstring-quotes = """
+
+# for snakemake pipeline
+builtins = snakemake
```
That's interesting! What is this line doing? Make flake8 check snakemake files specifically, or ignore them? Would be nice to add a comment. The reason is that I'm gonna use Snakemake along with flake8 in my projects too, so would be useful to add something on that into the team-docs repo as well.

mumichae replied:
I added some explanation in a comment. Basically, snakemake provides a variable called `snakemake`, when a script is provided in the rule. That way, the script has access to all the objects defined in the Snakefile and its rule. Flake8 does not recognise this, so I specified for it to ignore the `snakemake` variable.

vladsavelyev commented on README.md:
```diff
@@ -1 +1,84 @@
-# 3utr-variants
\ No newline at end of file
+# 3' UTR-variants
```
Nice and clean README :)

vladsavelyev commented on scripts/convert_bed.py:
```diff
@@ -0,0 +1,36 @@
+def convert_chr_style(
+    interval: pybedtools.Interval, chr_style: str
+) -> pybedtools.Interval:
+    if chr_style == '' and 'chr' in interval.chrom:
```
A bit nitpicky, but it would be nice to add a docstring to clarify the purpose of the function. It's pretty clear from the code in thise case, but it's a good practice anyway.

vladsavelyev commented on scripts/convert_bed.py:
```diff
@@ -0,0 +1,36 @@
+import pybedtools
```
pylint requires to add docstrings on top of every non-empty python file, which I agree is a good practice. It's not clear from the script name what kind of conversion it does. I'd add docstrings to every snakemake file as well.

vladsavelyev commented on scripts/extract_ployadb.py:
```diff
@@ -0,0 +1,123 @@
+"""Extract PolyA sites and cis-elements from PolyA_DB3"""
```
Typo in the name: `extract_ployadb.py ` -> `extract_polyadb.py ` :)

vladsavelyev commented:
I'm running the workflow with the default config, and it misses inputs:

```
$ snakemake -j1 -p
Workflow defines that rule genomepy is eligible for caching between workflows (use the --cache argument to enable this).
Building DAG of jobs...
WorkflowError:
InputFunctionException (rule convert_bedfile, line 6, /Users/vlad/git/populationgenomics/utr-variants/rules/evaluation.smk):
    Error:
      KeyError: 'PolyADB_hexamers_local'
    Wildcards:
      variant_subset=PolyADB_hexamers_local
    Traceback:
      File "/Users/vlad/git/populationgenomics/utr-variants/rules/evaluation.smk", line 7, in <lambda>
MissingInputException: Missing input files for rule extract_PolyA_DB:
data/annotations/PolyA_DB3/human.PAS.txt
```

Just suggesting that you could add rules into `download_data.smk` to automatically download the files from the `databases` section in the config

vladsavelyev commented on scripts/prepare_gnomAD.py:
```diff
@@ -0,0 +1,231 @@
+
+def get_worst_consequence_with_non_coding(ht):
+    """
+    copied and adapted from gnomad_lof
```
Would be nice to add a lint to the original to get some context. 
The function is hard to read, so if you could add docstrings and comments, and clarify which parts have you changed from the gnomad_lof codebase, that would help a lot to review it!

vladsavelyev commented on scripts/prepare_gnomAD.py:
```diff
@@ -0,0 +1,231 @@
+
+def prepare_ht(ht, trimer: bool = False):
+    if trimer:
+
```
Redundant empty line

vladsavelyev commented on scripts/prepare_gnomAD.py:
```diff
@@ -0,0 +1,231 @@
+    )
+
+
+def prepare_ht(ht, trimer: bool = False):
```
A docstring with the purpose of the function would be fantastic! And ideally, rename the function to a less generic name.

mumichae replied:
I copied the function from `gnomad_lof` directly. Added file and line number of the original in the docstring

vladsavelyev commented on scripts/prepare_gnomAD.py:
```diff
@@ -0,0 +1,231 @@
+def collapse_strand(
+    ht: Union[hl.Table, hl.MatrixTable]
+) -> Union[hl.Table, hl.MatrixTable]:
+    def reverse_complement_bases(
```
A docstring with the purpose! :)

vladsavelyev commented on scripts/maps_gcp.py:
```diff
@@ -0,0 +1,52 @@
+if __name__ == '__main__':
+    import argparse
+
+    parser = argparse.ArgumentParser(description='Compute MAPS score')
```
I suggest to add `help` parameters to the `add_argument` functions. Would improve the readability, otherwise it's hard to track down what does e.g. `context_ht` contain.

mumichae replied:
Added some help statements, however I'm not completely sure how exactly the `context_ht` is defined and what it contains. I'm trying to contact the authors/maintainers of the hail implementation of the MAPS score, so I'll hopefully know soon

vladsavelyev commented on scripts/maps_score.py:
```diff
@@ -0,0 +1,198 @@
+    impose_high_af_cutoff_here: bool = False,
+):
+    """
+    Count variants by context, ref, alt, methylation_level
```
The readability would improve if you add some detail on each function parameter and the return value. Something like here: https://github.com/populationgenomics/cpg-qc/blob/main/cpg_qc/hard_filtering.py#L28-L50

Also, don't forget to add type hints to the return value as well (e.g. `-> hl.Table`)

The function name is quite generic, so it would be nice to clarify the purpose of counting and how the output table is structured.

mumichae replied:
This function is from https://github.com/macarthur-lab/gnomad_lof, and I just copied it here to make sure the pipeline can run by itself. I forgot to add links to where I copied the function(s) from, so now they are included in the docstring. I would try to modify as little as possible with these functions, or just reuse their implementation entirely, however I'm not quite sure how to do that yet, hence the copy-paste solution

vladsavelyev replied:
I think one way would be to add gnomad_lof as a [git submodule](https://git-scm.com/book/en/v2/Git-Tools-Submodules):

```
git submodule add https://github.com/macarthur-lab/gnomad_lof
```

Then add it to the PYTHONPATH internally, and import them as python modules in `maps_score.py`. For `hailctl dataproc` submission, you'd have to append the gnomad_lof files that you are using into the `--pyfiles` command as well.

Though I don't see anything criminal in copying after you added the pointers to the original :)

mumichae replied:
Submodules are actually a pretty good solution to my problem, didn't think about them! I might consider them in a separate PR

vladsavelyev commented on scripts/maps_score.py:
```diff
@@ -0,0 +1,198 @@
+    singleton_expression: hl.expr.BooleanExpression = None,
+    skip_worst_csq: bool = False,
+) -> hl.Table:
+    if not skip_worst_csq:
```
A docstring would be useful here as well. Additionally, I'd comment each step in the function body, unless it follows some standard flow of calculating this MAPS score, in which case I'd put a link to the corresponding publication.

mumichae replied:
I added the github file & line number of this function. Hope that'll be enough

vladsavelyev commented:
It's really impressive how quickly you got familiar with Hail, and now using it for very complicated stuff. Really good job.

Also, appreciate your Snakemake skills, I learned a few things from your code.

However, to avoid the standard destiny of most scientific code - becoming abandoned after the author moves on to another project - I suggest focusing on improving the readability and the code design. E.g. add docstrings and comments; wherever reasonable, split into smaller and pure functions that do one thing; make meaningful variable/function names; add pointers to resources/publications that can explain code pieces. Try to imagine yourself looking at your code in 3 years: how quickly will you understand every piece of it?

But I should say you already made a great job in that direction by organising the code with Snakemake; by writing a clean readme; by providing the environment.yml file; by extracting paths into the config; by writing a very clean Python in general. Also nice that it supports both local and GCP runs. So don't worry too much if you are very time-restricted.

vladsavelyev reviewed (changes requested):
Submitting "Request changes" mainly because I wasn't able to run it yet (having some rules or docs on how to prepare the ref data would be helpful!)

Commit:
update lint configurations
add comment on injected snakemake variable
ignore line length check by pylint

Commit:
added more docstrings to scripts

Commit:
changed output files and final target handling

Commit:
include PolyA_DB and Gencode download

Commit:
Add docstrings for Snakefiles

Commit:
Add explanation of file structure

mumichae commented:
Thanks for the comments!
I went through most of them and added more documentation. However, I left gnomAD and MAPS related code undocumented, as it was mainly copied from the https://github.com/macarthur-lab/gnomad_lof repository. I had forgotten to add the references in the code, but they should now be in the docstrings of each function I copied. Ideally, I'd opt for a better way of code reuse, but haven't yet figured out how to do that. Copying the functions to the pipeline was an easy way to ensure that the pipeline is self-contained.

I also added downloads rules for the input databases (Gencode & PolyA_DB3), so you should be able to run it on your local machine now. Note, that for the subset I defined in the config, MAPS fails on the PAS hexamers. This is likely because the number of variants in those intervals is too small for a proper model fit.

Commit:
Added description of MAPS

vladsavelyev commented on rules/extract_UTR_features.smk:
```diff
@@ -16,8 +16,10 @@ rule extract_Gencode_PAS:
-        db=config["databases"]["PolyA_DB"]["file"],
-        fasta=expand(rules.genomepy.output[0], assembly=config['assembly_ucsc'])
+        db=rules.download_PolyA_DB.output,  # config["databases"]["PolyA_DB"]["file"],
+        fasta=ancient(
```
Nice! Another neat snakemake function that I didn't know :)

vladsavelyev commented on README.md:
```diff
@@ -9,9 +9,28 @@ This project involves:
+
+TODO: explain MAPS etc.
+
+## Repository Files
```
That's cool. Will really help future contributors to dive into the codebase.

vladsavelyev commented on README.md:
```diff
@@ -11,7 +11,8 @@ This project involves:
 
-TODO: explain MAPS etc.
+The different 3' UTR variants are evaluated by the MAPS (mutability-adjusted proportion of singletons) score to measure the overall depletion of singletons, as described in the [gnomAD flagship paper](https://doi.org/10.1038/s41586-020-2308-7).
+A higher MAPS score (compared to a baseline) suggests that there is a stronger evolutionary selection against the variant group.
```
Love this, short and precise.

vladsavelyev commented:
The update looks great, the docstrings really change the readability a lot.

vladsavelyev commented:
All good for this PR, approved and merged! Thanks a lot for addressing the requests.
